carball/analysis/events/carry_detection.py

```diff
@@ -0,0 +1,187 @@
+import logging
+from typing import Tuple, List
+
+import pandas as pd
+
+from carball.analysis.constants.field_constants import BALL_SIZE
+from carball.generated.api import game_pb2
+from carball.generated.api.player_pb2 import Player
+
+logger = logging.getLogger(__name__)
+
+class CarryData:
+    def __init__(self, carry_frames: pd.DataFrame, start_frames: List[int], end_frames: List[int]):
+        self.carry_frames = carry_frames
+        self.end_frames = end_frames
+        self.start_frames = start_frames
+
+
+class CarryDetection:
+
+    def filter_frames(self, data_frame: pd.DataFrame) -> CarryData:
+        valid_frames = data_frame[(data_frame.ball.pos_z > (BALL_SIZE + 5)) & (data_frame.ball.pos_z < 500)]
+        return CarryData(valid_frames, *self.creat_start_end_frames(valid_frames))
+
+    def create_carry_events(self, carry_data: CarryData, player: Player, proto_game: game_pb2.Game):
+        """
+        Gets all of the carry events and adds them to the protobuf for each individual player.
+        """
+        player_carry_data, xy_distance, player_frames = self.player_close_frames(carry_data.carry_frames,
+                                                                            carry_data.carry_frames[player.name])
+
+        modified_carry_data = self.expand_carries(carry_data, player_carry_data, player, proto_game)
+
+        self.add_carry_events(modified_carry_data, xy_distance, player_frames, player, proto_game)
+
+    def player_close_frames(self, valid_frames: pd.DataFrame,
+                            player_frames: pd.DataFrame) -> Tuple[CarryData, pd.DataFrame, pd.DataFrame]:
+        """
+        Filters out all frames that are not close enough to the player to be a valid carry.
+        Filters out dribbles that are too short
+        """
+
+        xy_distance = ((player_frames.pos_x - valid_frames.ball.pos_x) ** 2 +
+                       (player_frames.pos_y - valid_frames.ball.pos_y) ** 2) ** 0.5
+        carry_frames = valid_frames[(xy_distance < BALL_SIZE) & (player_frames.pos_z < valid_frames.ball.pos_z)]
+
+        player_carry_data = CarryData(valid_frames, *self.creat_start_end_frames(carry_frames))
+        return player_carry_data, xy_distance, player_frames
+        # look up hits.
+        # any hits by the same player within a continous set of valid frames should count as carrys
+
+    def merge_dribbles(self, carry_data: CarryData, player_carry_data: CarryData):
+        """
+        Merges separated dribbles that are still within the valid dribble set.
+        """
+        start_frames = player_carry_data.start_frames
+        end_frames = player_carry_data.end_frames
+
+        merged_start = []
+        merged_end = []
+
+        previous_start = -1
+        previous_end = 1
+
+        for frame_index in range(len(start_frames) - 1):
```

Comment:
There's a lot going on here with not much explanation. The docstring doesn't help much as it doesn't say what a valid dribble set is or what separated dribbles are. 
It's also hard to follow the logic within here (as I have no clue how it's trying to achieve what I think it's trying to achieve).